File: THCS/2022_2023/LamDong_2223/Cau2/CAU2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <utility> // cho std::pair
#include <cmath>   // cho sqrt

using namespace std;

// Hàm kiểm tra số nguyên tố
bool isPrime(long long num) {
    if (num < 2) return false;
    for (long long i = 2; i <= sqrt(num); i++) {
        if (num % i == 0) return false;
    }
    return true;
}

// Hàm kiểm tra giả thuyết Goldbach
void checkGoldbachConjecture(int n, const vector<long long>& numbers, ofstream& outputFile) {
    for (long long k : numbers) {
        pair<long long, long long> bestPair(0, 0);
        bool found = false;

        // Tìm cặp số nguyên tố i và j sao cho i + j = k
        for (long long i = 2; i <= k / 2; i++) {
            long long j = k - i;
            if (isPrime(i) && isPrime(j)) {
                // Cập nhật cặp số nếu là cặp tốt hơn
                if (abs(i - j) > abs(bestPair.first - bestPair.second)) {
                    bestPair = make_pair(i, j);
                }
                found = true;
            }
        }

        // Ghi kết quả vào file
        if (found) {
            outputFile << min(bestPair.first, bestPair.second) << " " << max(bestPair.first, bestPair.second) << endl;
        } else {
            outputFile << "NO" << endl;
        }
    }
}

int main() {
    ifstream inputFile("CAU2.INP");
    ofstream outputFile("CAU2.OUT");

    int n;
    inputFile >> n; // Đọc số lượng test
    vector<long long> numbers(n);

    // Đọc các số chẵn
    for (int i = 0; i < n; i++) {
        inputFile >> numbers[i];
    }

    // Kiểm tra giả thuyết Goldbach
    checkGoldbachConjecture(n, numbers, outputFile);

    inputFile.close();
    outputFile.close();
    return 0;
}

File: THCS/2022_2023/LamDong_2223/Cau2/CAU2.py
def is_prime(num):
    """Kiểm tra xem số num có phải là số nguyên tố hay không."""
    if num < 2:
        return False
    for i in range(2, int(num**0.5) + 1):
        if num % i == 0:
            return False
    return True

def goldbach_conjecture(n, numbers):
    results = []
    for k in numbers:
        found = False
        best_pair = (0, 0)
        # Tìm cặp số nguyên tố có tổng bằng k
        for i in range(2, k // 2 + 1):
            j = k - i
            if is_prime(i) and is_prime(j):
                # Cập nhật cặp số nếu là cặp tốt hơn
                if abs(i - j) > abs(best_pair[0] - best_pair[1]):
                    best_pair = (i, j)
                found = True
        
        if found:
            results.append(f"{min(best_pair)} {max(best_pair)}")
        else:
            results.append("NO")
    return results

# Đọc dữ liệu từ file
with open("CAU2.INP", "r") as f:
    n = int(f.readline().strip())
    numbers = [int(f.readline().strip()) for _ in range(n)]

# Kiểm tra giả thuyết Goldbach
results = goldbach_conjecture(n, numbers)

# Ghi kết quả vào file
with open("CAU2.OUT", "w") as f:
    f.write("\n".join(results) + "\n")
